src/libmeasurement_kit/ooni/whatsapp.cpp

```diff
@@ -0,0 +1,575 @@
+        }
+    }
+
+    int rem_bits = pre_bits % 8;
+    if (    (ip1[i] >> (8-rem_bits)) != \
+            (ip2[i] >> (8-rem_bits))) {
+        return false;
+    }
+    return true;
+}
+
+ErrorOr<bool> ip_in_net(std::string ip1, std::string ip_w_mask) {
+    auto ip1bs = ip_to_bytes(ip1);
+    for (auto b : ip1bs) {
+    }
+    auto ip2 = mk::split<std::vector<std::string>>(ip_w_mask, "/");
+    auto ip2bs = ip_to_bytes(ip2[0]);
+    for (auto b : ip2bs) {
+    }
+    auto pre_bits = std::stoi(ip2[1]);
+    return same_pre(ip1bs, ip2bs, pre_bits);
+}
+
+bool ip_in_nets(std::string ip, std::vector<std::string> nets) {
+    for (auto net : nets) {
+        auto res = ip_in_net(ip, net);
+        if (!!res && (res.as_value() == true)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+// XXX not even beginning to fill in entry
+static void tcp_many(std::vector<std::string> ips,
+                     Var<Entry> entry,
+                     Settings options,
+                     Var<Reactor> reactor,
+                     Var<Logger> logger,
+                     Callback<Error> cb) {
+    // two ports per IP
+    int ips_count = ips.size() * 2;
+    Var<int> ips_tested(new int(0));
+    if (ips_count == *ips_tested) {
+        cb(NoError());
+        return;
+    }
+
+    auto tcp_cb = [=](std::string ip, int port) {
+        return [=](Error err, Var<net::Transport> txp) {
+            bool close_txp = true; // if connected, we must disconnect
+            Entry result = {
+                {"ip", ip},
+                {"port", port},
+                {"status",
+                    {{"success", nullptr},
+                    {"failure", nullptr}}},
+            };
+            if (!!err) {
+                logger->info("tcp failure to %s:%d", ip.c_str(), port);
+                result["status"]["success"] = false;
+                result["status"]["failure"] = err.as_ooni_error();
+                close_txp = false;
+            } else {
+                logger->info("tcp success to %s:%d", ip.c_str(), port);
+                result["status"]["success"] = true;
+                result["status"]["failure"] = false;
+            }
+            (*entry)["tcp_connect"].push_back(result);
+            *ips_tested += 1;
+            if (ips_count == *ips_tested) {
+                if (close_txp == true) {
+                    txp->close([=] { cb(NoError()); });
+                } else {
+                    cb(NoError());
+                }
+            } else {
+                if (close_txp == true) {
+                    // XXX optimistic closure
+                    txp->close([=] {});
+                }
+            }
+        };
+    };
+
+    for (auto const& ip : ips) {
+        // XXX hardcoded
+        std::vector<int> ports {443, 5222};
+        for (auto const& port : ports) {
+            options["host"] = ip;
+            options["port"] = port;
+            options["net/timeout"] = 10.0; //XXX check if set upstream?
```

Comment:
The timeout is set by default. I'd initialize the local options variable from the options of the test.

Reply:
acked (and I can't reply to the other options comment for some reason)